File: semantics/semantics.js
// JavaScript: dynamic typing, lexical scope, and closure state
function makeAccumulator(start) {
    let total = start;
    return function add(value) {
        total += value;
        return total;
    };
}

const acc = makeAccumulator(10);
console.log(acc(5));
console.log(acc("2")); // Numeric 15 becomes string "152" through coercion.

File: semantics/semantics.py
# Python: dynamic typing, lexical scope, and nonlocal closure state
def make_accumulator(start):
    total = start
    def add(value):
        nonlocal total
        total += value
        return total
    return add

acc = make_accumulator(10)
print(acc(5))
print(acc(2.5))  # The closure accepts a float and total becomes a float.
